src/neo/SmartContract/Native/NameService.cs

```diff
@@ -0,0 +1,290 @@
+        private void SetPrice(ApplicationEngine engine, long price)
+        {
+            if (price <= 0 || price > 10000_00000000) throw new ArgumentOutOfRangeException(nameof(price));
+            if (!CheckCommittee(engine)) throw new InvalidOperationException();
+            engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_DomainPrice)).Set(price);
+        }
+
+        [ContractMethod(0_01000000, CallFlags.ReadStates)]
+        public long GetPrice(StoreView snapshot)
+        {
+            return (long)(BigInteger)snapshot.Storages[CreateStorageKey(Prefix_DomainPrice)];
+        }
+
+        [ContractMethod(0_01000000, CallFlags.ReadStates)]
+        public bool IsAvailable(StoreView snapshot, string name)
+        {
+            if (!nameRegex.IsMatch(name)) throw new ArgumentException(null, nameof(name));
+            string[] names = name.Split('.');
+            if (names.Length != 2) throw new ArgumentException(null, nameof(name));
+            byte[] hash = GetKey(Utility.StrictUTF8.GetBytes(name));
+            if (snapshot.Storages.TryGet(CreateStorageKey(Prefix_Token).Add(hash)) is not null) return false;
+            StringList roots = snapshot.Storages.TryGet(CreateStorageKey(Prefix_Roots))?.GetInteroperable<StringList>();
+            if (roots is null || roots.BinarySearch(names[1]) < 0) throw new InvalidOperationException();
+            return true;
+        }
+
+        [ContractMethod(0_01000000, CallFlags.WriteStates)]
+        private bool Register(ApplicationEngine engine, string name, UInt160 owner)
+        {
+            if (!nameRegex.IsMatch(name)) throw new ArgumentException(null, nameof(name));
+            string[] names = name.Split('.');
+            if (names.Length != 2) throw new ArgumentException(null, nameof(name));
+            if (!engine.CheckWitnessInternal(owner)) throw new InvalidOperationException();
+            byte[] hash = GetKey(Utility.StrictUTF8.GetBytes(name));
+            if (engine.Snapshot.Storages.TryGet(CreateStorageKey(Prefix_Token).Add(hash)) is not null) return false;
+            StringList roots = engine.Snapshot.Storages.TryGet(CreateStorageKey(Prefix_Roots))?.GetInteroperable<StringList>();
+            if (roots is null || roots.BinarySearch(names[1]) < 0) throw new InvalidOperationException();
+            engine.AddGas(GetPrice(engine.Snapshot));
+            NameState state = new NameState
+            {
+                Owner = owner,
+                Name = name,
+                Description = "",
+                Expiration = (uint)(engine.Snapshot.PersistingBlock.Timestamp / 1000) + OneYear
+            };
+            Mint(engine, state);
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_Expiration).AddBigEndian(state.Expiration).Add(hash), new StorageItem(new byte[] { 0 }));
+            return true;
+        }
+
+        [ContractMethod(0, CallFlags.ReadStates)]
+        public uint ExpireOn(StoreView snapshot, string name)
```

Comment:
The `properties` method will return this field.

Reply:
Done